shenyu-plugin/shenyu-plugin-request/src/main/java/org/apache/shenyu/plugin/request/RequestPlugin.java

```diff
@@ -0,0 +1,215 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * The RequestPlugin.
+ */
+@Slf4j
+public class RequestPlugin extends AbstractShenyuPlugin {
+
+    @Override
+    protected Mono<Void> doExecute(final ServerWebExchange exchange, final ShenyuPluginChain chain, final SelectorData selector, final RuleData rule) {
+        RequestHandle requestHandle = RequestRuleHandleCache.getInstance().obtainHandle(CacheKeyUtils.INST.getKey(rule));
+        if (Objects.isNull(requestHandle)) {
+            log.error("uri request rule can not configuration.");
+            return chain.execute(exchange);
+        }
+        ServerHttpRequest request = exchange.getRequest();
+        ServerWebExchange modifiedExchange = exchange.mutate()
+                .request(originalRequest -> originalRequest.uri(
+                        UriComponentsBuilder.fromUri(exchange.getRequest()
+                                .getURI())
+                                .replaceQueryParams(getQueryParams(request, requestHandle))
+                                .build()
+                                .toUri()
+                        ).headers(httpHeaders -> httpHeaders.addAll(getHeaders(request, requestHandle)))
+                ).build();
+        return chain.execute(modifiedExchange);
+    }
+
+    @Override
+    public int getOrder() {
+        return PluginEnum.REQUEST.getCode();
+    }
+
+    @Override
+    public String named() {
+        return PluginEnum.REQUEST.getName();
+    }
+
+    /**
+     * getHeaders.
+     *
+     * @param request       serverHttpRequest
+     * @param requestHandle requestHandle
+     * @return new headers
+     */
+    public HttpHeaders getHeaders(final ServerHttpRequest request, final RequestHandle requestHandle) {
+        RequestHandle.ShenyuRequestHeader shenyuReqHeader = requestHandle.getHeader();
+        HttpHeaders headers = HttpHeaders.writableHttpHeaders(request.getHeaders());
+        if (Objects.isNull(shenyuReqHeader)) {
+            return headers;
+        }
+        if (Objects.nonNull(shenyuReqHeader.getAddHeaders())
+                && MapUtils.isNotEmpty(shenyuReqHeader.getAddHeaders())) {
+            shenyuReqHeader.getAddHeaders().entrySet().forEach(s -> this.fillHeader(s, headers));
+        }
+        if (Objects.nonNull(shenyuReqHeader.getSetHeaders())
+                && MapUtils.isNotEmpty(shenyuReqHeader.getSetHeaders())) {
+            shenyuReqHeader.getSetHeaders().entrySet().forEach(s -> this.fillHeader(s, headers));
+        }
+        if (Objects.nonNull(shenyuReqHeader.getReplaceHeaderKeys())
+                && MapUtils.isNotEmpty(shenyuReqHeader.getReplaceHeaderKeys())) {
+            shenyuReqHeader.getReplaceHeaderKeys().entrySet().forEach(s -> this.replaceHeaderKey(s, headers));
+        }
+        if (Objects.nonNull(shenyuReqHeader.getRemoveHeaderKeys())
+                && CollectionUtils.isNotEmpty(shenyuReqHeader.getRemoveHeaderKeys())) {
```

Comment:
`CollectionUtils.isNotEmpty` is already included null check. Should we remove `Objects.nonNull`? By the way, you can also check the `MapUtils`. : )

Reply:
good idea,thx.